### PR TITLE
refactor(amf): re-aligned smfcontext structure parameters

### PR DESCRIPTION
**Pull request by ganeshg87**
Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

## Summary

smf_context_t structure parameters re-aligned, removed unused members

## Test Plan

- verified Registration, PDU Session Establishment and Release, Deregistration using UERANSIM
image
[mme_reg_14_Dec_21_R1.log](https://github.com/magma/magma/files/7711977/mme_reg_14_Dec_21_R1.log)

- verified test oai
image

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h [hunk 1]
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
...
-  MessageTypeMsg message_type;
-  IntegrityProtMaxDataRateMsg integrity_prot_max_data_rate;
-  PDUSessionTypeMsg pdu_session_type;
-  SSCModeMsg ssc_mode;
+  uint8_t pdu_session_id;
+  uint8_t pti;
+  uint8_t message_type;
+  uint8_t max_uplink;
+  uint8_t max_downlink;
+  uint32_t pdu_session_type;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: Shouldn't this be an enum type?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h [hunk 2]
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
...
+  uint32_t ssc_mode;
 } smf_proc_data_t;
 
+typedef struct session_ambr_s {
+  uint8_t dl_ambr_unit;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: Enum?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h [hunk 3]
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
...
+  uint16_t dl_session_ambr;
+  uint8_t ul_ambr_unit;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: Enum?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h [hunk 4]
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
...
 typedef struct smf_proc_data_s {
-  PDUSessionIdentityMsg pdu_session_identity;
-  PTIMsg pti;
-  MessageTypeMsg message_type;
-  IntegrityProtMaxDataRateMsg integrity_prot_max_data_rate;
-  PDUSessionTypeMsg pdu_session_type;
-  SSCModeMsg ssc_mode;
+  uint8_t pdu_session_id;
+  uint8_t pti;
+  uint8_t message_type;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: enum?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h [hunk 5]
```diff
@@ -242,42 +242,46 @@ typedef struct teid_upf_gnb_s {
...
-  uint8_t sst;
-  uint8_t sd[SD_LENGTH];
+  s_nssai_t requested_nssai;
 
-  // Request to gnb on PDU establisment request
-  pdu_session_resource_setup_req_t pdu_resource_setup_req;
-  pdu_session_resource_setup_rsp_t pdu_resource_setup_rsp;
-  pdu_session_resource_to_release_list pdu_resource_release_req;
-  pdu_session_resource_to_release_list pdu_resource_release_rsp;
+  qos_flow_request_list_t subscribebed_qos_profile;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]:
```suggestion
  qos_flow_request_list_t subscribed_qos_profile;
```

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 1 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: it was an Existing code.
- I think mainly to pass SMF GRPC message directly without converting to the corresponding enum.
- changing to Enum requires more amf changes for message_type, pdu_session_type, dl_ambr_unit.

@ssanadhya, Please suggest if needed we will take it as a separate activity.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 5 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 1 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h:250]: it was an Existing code.

I think mainly to pass SMF GRPC message directly without converting to the corresponding enum.
changing to Enum requires more amf changes for message_type, pdu_session_type, dl_ambr_unit.
@ssanadhya, Please suggest if needed we will take it as a separate activity.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 5 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 1 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: @ganeshg87 , Pdu Session Type is specified as enum in [session_manager.proto](https://github.com/magma/magma/blob/master/lte/protos/session_manager.proto#L695) and the AMF GRPC servicer interprets it as [pdu_session_type_t](https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/grpc_service/AmfServiceImpl.cpp#L83), which is also an [enum](https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/include/n11_messages_types.h#L104). Am I missing something here?

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: **session_manager**.proto :
// PduSessionType
enum PduSessionType {
   IPV4 = 0;
   IPV6 = 1;
   IPV4IPV6 = 2;
   UNSTRUCTURED = 3;
}
//nas message enum values as per spec(501)
enum class M5GPduSessionType : uint8_t {
  IPV4         = 0b001,
  IPV6         = 0b010,
  IPV4V6       = 0b011,
  UNSTRUCTURED = 0b100,
  ETHERNET     = 0b101,
};

both enum values are mis-match. both enums should follow the same values. 
If I replace enum defined in sesssion_manager.proto,  Need to validate sessiond & amf interworking behavior.

I accept your review comment, these all these should define as enum

Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: @ganeshg87 , what's the plan to revise this PR?

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: @ssanadhya, I have addressed review comments related to Enum change and updated all changes in current PR.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 4 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 3 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h (hunk 2 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h]: Addressed review comment

**Comment by ganeshg87** [comment]
Hi @ssanadhya, 
Refactoring Existing Code will consume additional time and testing efforts. It will be difficult manage to  Stateless Feature deliverable timelines. We would need your support to raise separate tasks if changes will take more time.
cc: @panyogesh

**Comment by ganeshg87** [comment]
verified basic Registration, PDU Session Setup, and Release, Deregistration using UERANSIM.
[mme_reg_pdu_rel_derg_20.log](https://github.com/magma/magma/files/7906872/mme_reg_pdu_rel_derg_20.log)

image
test_oai_20_jan

**Comment by ganeshg87** [comment]
@ssanadhya,
 I have addressed Enum-related changes, Please provide your feedback.

Thank you
Ganesh

File: lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp [hunk 6]
```diff
@@ -307,10 +311,11 @@ int nas5g_message_encode(
       }
     }
     OAILOG_INFO(
-        LOG_AMF_APP, "Encoded msg(nas5g) id: [%x]-name [%s]",
+        LOG_AMF_APP, "Encoded msg(nas5g) id: [%lu]-name [%s]",
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp]: this should be `%u`?

File: lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp [hunk 7]
```diff
@@ -320,9 +325,11 @@ int nas5g_message_encode(
         _nas5g_message_plain_encode(buffer, &msg->header, &msg->plain, length);
 
     OAILOG_INFO(
-        LOG_AMF_APP, "Encoded msg(nas5g) id: [%x]-name [%s]",
+        LOG_AMF_APP, "Encoded msg(nas5g) id: [%lu]-name [%s]",
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp]: same comment here regarding `%u` formatting

**Review by ardzoht — APPROVED** [review]
lgtm, left 2 minor comments, can be merged with @ssanadhya's approval

File: lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp (hunk 6 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp]: @ardzoht, 

the actual intention was to print hex value for message_type.I have reverted back to hex print.

000465 Mon Jan 24 11:19:10 2022 7F76CE0E1700 INFO  AMF-AP tasks/amf/nas5g_message.cpp     :0327    Encoded msg(nas5g) id: [**56**]-name [AUTH_REQUEST].

File: lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp (hunk 7 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp]: the actual intention was to print hex value for message_type.I have reverted back to hex print.

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp [hunk 8]
```diff
@@ -175,59 +175,59 @@ static int amf_as_establish_req(amf_as_establish_t* msg, int* amf_cause) {
...
+  switch (static_cast<M5GMessageType>(amf_msg->header.message_type)) {
+    case M5GMessageType::REG_REQUEST:
       memcpy(&originating_tai, &msg->tai, sizeof(originating_tai));
       rc = amf_handle_registration_request(
           msg->ue_id, &originating_tai, &msg->ecgi,
           &amf_msg->msg.registrationrequestmsg, msg->is_initial,
           msg->is_amf_ctx_new, *amf_cause, decode_status);
       break;
-    case M5G_SERVICE_REQUEST:  // SERVICE_REQUEST:
+    case M5GMessageType::M5G_SERVICE_REQUEST:  // SERVICE_REQUEST:
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]:
```suggestion
    case M5GMessageType::M5G_SERVICE_REQUEST: 
```

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp [hunk 9]
```diff
@@ -1056,15 +1065,17 @@ int construct_pdu_session_reject_dl_req(
...
     msg->header.security_header_type =
         SECURITY_HEADER_TYPE_INTEGRITY_PROTECTED_CYPHERED;
     dlmsg = &msg->security_protected.plain.amf.msg.downlinknas5gtransport;
   } else {
     msg->plain.amf.header.extended_protocol_discriminator =
         M5G_MOBILITY_MANAGEMENT_MESSAGES;
-    msg->plain.amf.header.message_type = DLNASTRANSPORT;
-    msg->header.security_header_type   = SECURITY_HEADER_TYPE_NOT_PROTECTED;
+    msg->plain.amf.header.message_type =
+        static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: What's the need to cast from "M5GMessageType" as "M5GMessageType"?

File: lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp [hunk 10]
```diff
@@ -185,13 +185,17 @@ int nas5g_message_decode(
...
-      get_message_type_str(msg->plain.amf.header.message_type).c_str(), bytes,
-      uint8_to_hex_string(buffer, bytes).c_str());
+      get_message_type_str(
+          static_cast<uint8_t>(msg->plain.amf.header.message_type))
+          .c_str(),
+      bytes, uint8_to_hex_string(buffer, bytes).c_str());
 
   OAILOG_INFO(
-      LOG_AMF_APP, "Decoded msg(nas5g) id: [%x]-name [%s]",
+      LOG_AMF_APP, "Decoded msg(nas5g) id: [%lu]-name [%s]",
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp]: Please address these lint comments.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 11]
```diff
@@ -707,12 +716,14 @@ TEST(test_dlnastransport, test_dlnastransport) {
   msg.plain.amf.header.extended_protocol_discriminator =
       M5G_MOBILITY_MANAGEMENT_MESSAGES;
   msg.header.extended_protocol_discriminator = M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.plain.amf.header.message_type          = DLNASTRANSPORT;
+  msg.plain.amf.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: What's the need to cast from "M5GMessageType" to "M5GMessageType"?

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 12]
```diff
@@ -707,12 +716,14 @@ TEST(test_dlnastransport, test_dlnastransport) {
...
   msg.header.security_header_type = SECURITY_HEADER_TYPE_NOT_PROTECTED;
   // SECURITY_HEADER_TYPE_INTEGRITY_PROTECTED_CYPHERED;
   msg.header.extended_protocol_discriminator = M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.header.message_type                    = DLNASTRANSPORT;
-  msg.header.sequence_number                 = 1;
+  msg.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: Same comment as above for casting from M5GMessageType to M5GMessageType, please check in rest of the code as well.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 13]
```diff
@@ -761,7 +759,8 @@ int amf_app_handle_pdu_session_accept(
   // Message construction for PDU Establishment Accept
   msg.security_protected.plain.amf.header.extended_protocol_discriminator =
       M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.security_protected.plain.amf.header.message_type = DLNASTRANSPORT;
+  msg.security_protected.plain.amf.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Casting "M5GMessageType" to "M5GMessageType"

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 14]
```diff
@@ -160,7 +161,8 @@ int pdu_session_resource_release_request(
 
   msg.security_protected.plain.amf.header.extended_protocol_discriminator =
       M5G_MOBILITY_MANAGEMENT_MESSAGES;
-  msg.security_protected.plain.amf.header.message_type = DLNASTRANSPORT;
+  msg.security_protected.plain.amf.header.message_type =
+      static_cast<M5GMessageType>(M5GMessageType::DLNASTRANSPORT);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: Same comment as above

**Review by ssanadhya — APPROVED** [review]
Looks good, with minor nits. Please revise and then mark it as Ready2Merge. cc: @ardzoht

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp (hunk 8 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp (hunk 10 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/nas5g_message.cpp]: addressed review comment

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp (hunk 11 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: addressed review comment, typecast is not needed

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp (hunk 12 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 13 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp (hunk 14 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: addressed review comment

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp (hunk 9 above)
Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: addressed review comment

**Comment by ganeshg87** [comment]
> Looks good, with minor nits. Please revise and then mark it as Ready2Merge. cc: @ardzoht

@ssanadhya addressed review comments and marked Ready2Merge. cc: @ardzoht

**Pull request merged**